dali/operators/generic/roi_random_crop.cc

```diff
@@ -0,0 +1,207 @@
+  if (in_shape_arg_.IsDefined() || ws.NumInput() == 1) {
+    if (in_shape_arg_.IsDefined() && (ws.NumInput() == 1)) {
+      DALI_FAIL("``in_shape`` argument is incompatible with providing an input.")
+    }
+    if (in_shape_arg_.IsDefined()) {
+      in_shape_.resize(nsamples, ndim);
+      in_shape_arg_.Acquire(spec_, ws, nsamples, sh);
+      for (int s = 0; s < nsamples; s++) {
+        auto sample_sh = in_shape_.tensor_shape_span(s);
+        for (int d = 0; d < ndim; d++) {
+          sample_sh[d] = in_shape_arg_[s].data[d];
+        }
+      }
+    } else {
+      auto &in = ws.template InputRef<CPUBackend>(0);
+      in_shape_ = in.shape();
+    }
+
+    for (int s = 0; s < nsamples; s++) {
+      auto sample_sh = in_shape_.tensor_shape_span(s);
+      for (int d = 0; d < ndim; d++) {
+        DALI_ENFORCE(sample_sh[d] >= crop_shape_[s].data[d],
+                     "Cropping shape can't be bigger than the input shape.");
+      }
+      if (roi_shape_.IsDefined()) {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 &&
+                       sample_sh[d] >= (roi_start_[s].data[d] + roi_shape_[s].data[d]),
+                       "ROI can't be out of bounds.");
+        }
+      } else {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 && sample_sh[d] >= roi_end_[s].data[d],
+                       "ROI can't be out of bounds.");
+        }
+      }
+    }
+  }
+
+  output_desc.resize(1);
+  output_desc[0].shape = uniform_list_shape(nsamples, sh);
+  output_desc[0].type = TypeTable::GetTypeInfo(DALI_INT64);
+  return true;
+}
+
+void ROIRandomCropCPU::RunImpl(workspace_t<CPUBackend> &ws) {
+  auto &out_crop_start = ws.template OutputRef<CPUBackend>(0);
+  auto crop_start = view<int64_t, 1>(out_crop_start);
+
+  int nsamples = crop_start.shape.size();
+  int ndim = crop_start[0].shape[0];
+
+  auto& thread_pool = ws.GetThreadPool();
+
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    thread_pool.AddWork(
+      [&, ndim, sample_idx](int thread_id) {
+        int64_t* sample_sh = nullptr;
+        if (!in_shape_.empty())
+          sample_sh = in_shape_.tensor_shape_span(sample_idx).data();
+        for (int d = 0; d < ndim; d++) {
+          int64_t roi_extent = -1;
+          int64_t roi_start = roi_start_[sample_idx].data[d];
+          int64_t crop_extent = crop_shape_[sample_idx].data[d];
+          if (roi_end_.IsDefined()) {
+            roi_extent = roi_end_[sample_idx].data[d] - roi_start;
+          } else {
+            roi_extent = roi_shape_[sample_idx].data[d];
+          }
+
+          if (roi_extent == crop_extent) {
+            crop_start[sample_idx].data[d] = roi_start;
+          } else if (roi_extent > crop_extent) {
+            int64_t range_end = roi_start + roi_extent - crop_extent;
+            if (sample_sh)
+              range_end = std::min<int64_t>(sample_sh[d] - crop_extent, range_end);
```

Comment:
I think you check if ROI is inside shape already in setup. So this is not needed.